### ISSUE TITLE
GPX IO

### ISSUE THREAD

**Issue by xivk** [NetTopologySuite/NetTopologySuite#147]
Would it be something we could implement, a GPX reader/writer? User @HarelM is asking this in this OsmSharp issue:

https://github.com/OsmSharp/core/issues/21

I removed the GPX IO code from OsmSharp because I believe this belongs more in a library like NTS correct? Or do we full depend on what JTS has? 

Can we alternatively do something seperate like a NetTopologySuite.IO.Gpx project that lives in a seperate repo?

**Comment by FObermaier** [comment]
I'd appreciate GPX IO code.
We have a lot of supporting IO code that JTS does not have like e.g. reading postgis or spatialite geometry.

**Comment by HarelM** [comment]
I have used the following code and have no issues reading and writing GPX version 1.1:
https://github.com/IsraelHikingMap/Site/blob/master/IsraelHiking.API/Gpx/GpxTypes.xsd
Generated code from the above xsd can be found here:
https://github.com/IsraelHikingMap/Site/blob/master/IsraelHiking.API/Gpx/GpxTypes.cs
Using regular ```XmlSerializer``` and the above code does a perfect and easy job as far as I can tell...

I think a dedicated project just for these two files is an overkill, but I also think many DLLs (and projects) can lead to circular dependencies and hard to maintain code (also in terms of framework migrations etc). 
My suggestion is to combine all IO code into the same project unless every IO project has its own dependencies which will create a DLL which is too big and will have too many dependencies, in that case split it to smaller groups, but not a DLL per format.
My two cents anyway. your project - your decision.

**Comment by FObermaier** [comment]
> My suggestion is to combine all IO code into the same project unless every IO project has its own dependencies which will create a DLL which is too big and will have too many dependencies, in that case split it to smaller groups, but not a DLL per format.

You are right, as long as there are only framework dependencies we can put it all together in one project (NetTopologySuite.IO), as soon as we need to depend on third party libraries it has to be seperate.

**Comment by FObermaier** [comment]
@HarelM, I did incorporate your GPX IO code into NetTopologySuite.IO project.
I did have to make some modification so it would work with PCL, could you please review that.
Perhaps there need to be more modifcations for it to work with netStandard.

**Comment by HarelM** [comment]
Would be happy to! 
Can you please point me to the right branch/folder?

**Comment by FObermaier** [comment]
It is in the `Refactor_for_netStandard` branch.
The folder is `NetTopologySuite.IO/NetTopologySuite.IO/Gpx`

**Comment by HarelM** [comment]
While I think that for my personal use the ```ToBytes``` and ```ToGpx``` methods are extremely useful this might not be the right approach for this framework (note that I have written them and they are not generated from the xsd).
Maybe a more consistent approach will be to write two classes called ```GpxWriter``` and ```GpxReader``` that receive a stream, what do you think?
I currently have a lot of exception thrown from these methods when I try to desalinize bytes that are not a valid Gpx so the reader may have more code.
Keep up the good work!

**Comment by airbreather** [comment]
As a note, I did not include the GPX code in my own branch for fixing #110, so this is still open.

**Comment by HarelM** [comment]
Will it be possible to add it back? I have removed the code from my project in order to use the code that was moved here, so this will mean I won't be able use the newly created NuGet packages without re-importing the GPX code :-/

**Comment by airbreather** [comment]
> Will it be possible to add it back?

I'll take this on.  Hoping to have it done in a week or so.

**Comment by HarelM** [comment]
It's only one file...

**Comment by airbreather** [comment]
> It's only one file...

Two things...

1. Unlike what the `Refactor_for_NETStandard` branch did, my branch didn't immediately combine many of the NetTopologySuite.IO.* projects into a single project, so I'm thinking the change would probably be one of the following:
    1. Dump it into `NetTopologySuite.IO` knowing that that refactoring is coming "soon", or
    2. Make `NetTopologySuite.IO.GPX` knowing that we're going to keep this separation going forward.
2. I'd like at least some automated tests for this, using real-world sample data.

**Comment by HarelM** [comment]
I can provide any number of GPX files as you'd need.
I think (might not be a consensus) that there shouldn't be too many IO libraries.
Since this file type specifically doesn't need any other dependency I think it's safe to dump it in `NetTopologySuite.IO` for the time being as this is a new feature and should pave the way for future IO code.

**Comment by airbreather** [comment]
@HarelM: In your extension methods, why does `UpdateBounds` short-circuit when the bounds are already set?  Does it **need** to do this?

**Comment by airbreather** [comment]
Going even further into this... I really don't see the purpose of adding this to NTS in anywhere near its current state.  It doesn't even use GeoAPI or NetTopologySuite types: once a consumer deserializes it, the burden falls to them to use a completely disjointed API to access the geometry data stored within it.

Contrast with nearly* all the other `NetTopologySuite.IO.*` projects, which yield either simple `IGeometry` objects or, in the case of at least `NetTopologySuite.IO.GeoTools`, `NetTopologySuite.Features.Feature` objects that also have metadata on top of them.

**the only exception to this that I can find is `NetTopologySuite.IO.Streams.CloudStorage`.  I don't know why that project exists.  It exposes a single class that just looks like a thin wrapper around Azure.*

The rest of this is just commentary on the rest of the code, in case I'm wrong.

re: (de/)serialization extension methods, I don't think it makes sense to have the JSON `FeatureCollection` stuff like what was [over there](https://github.com/NetTopologySuite/NetTopologySuite/blob/77fd2ae8c932c7fccbf37b9e476bc029dcb0ac43/NetTopologySuite.IO/NetTopologySuite.IO/IO/Gpx/GpxSerializationExtensions.cs#L18-L42) (hidden behind a "never compile this" flag).

I also don't immediately think that the `gpxType` <--> `byte[]` methods are appropriate for NTS proper either.  IMO, the furthest we should go is expose one method that reads a single GPX element from an active `XmlReader` provided by the caller (which might be inside some outer list, backed by a remote `Stream`, encoded in some unusual encoding), and the `XmlWriter` equivalent for producing the same.

I came up with [this](https://gist.github.com/airbreather/317c3c8023305ce00cd00b84f921ac07), again just in case I'm wrong and something like this does in fact belong in NTS.

**Comment by HarelM** [comment]
See my comments above about what should be used from my code, I generally agree with you.
I also have some code to convert gpx to geojson I can share if you think it's relevant for this library.
When using in memory objects one should use geometry etc, but serialization of gpx requires the xml types in order to read and write the xml, I think it should be in the io project as do other formats...

**Comment by airbreather** [comment]
> When using in memory objects one should use geometry etc, but serialization of gpx requires the xml types in order to read and write the xml, I think it should be in the io project as do other formats...

This could be done without using XSD-to-C# and exposing those implementation details (which doesn't have a chance of working with .NET Standard 1.0, but maybe that's OK).

Instead, one possibility would be to just say:

```csharp
public static class Gpx
{
    // IGeometry might be embedded in some outer container that preserves metadata
    public static IGeometry Read(XElement gpxElement)
    {
        // ...
    }

    public static XElement Write(IGeometry geometry)
    {
        // ...
    }
```

> I also have some code to convert gpx to geojson I can share if you think it's relevant for this library.

That actually sounds like a really good idea.  From a quick Google search, the format looks like it hasn't been updated in over 13 years, so it might be worthwhile to just solve GPX by converting to GeoJSON and then leveraging everything from `NetTopologySuite.IO.GeoJSON`, or letting the caller do so themselves (maybe they just want to convert GPX to GeoJSON as a one-time thing and then move forward with the newer, more actively maintained format).

**Comment by airbreather** [comment]
> I think it should be in the io project as do other formats...

For the most part, I agree, but there are a few that I think should continue to be separate:

1. GeoJSON + TopoJSON (requires an extra package reference)
2. GeoTools (until #184 is resolved)
3. Oracle (requires lots of extra crap, doesn't even build with the main solution today)
4. SqlServer2008 (requires some stuff that .NET Standard 1.x doesn't support, plus extra packages)
5. Streams.CloudStorage (I don't even know why this exists)

**Comment by HarelM** [comment]
Here's the relevant conversion code, feel free to take it.
https://github.com/IsraelHikingMap/Site/blob/master/IsraelHiking.API/Converters/GpxGeoJsonConverter.cs
In order to write things to file that are relevant only to gpx one needs the gpx types, I don't see a way around it.
I can't relate to the statement that json is a dependency as I don't think there's an app that is written nowadays that does not use this package, but maybe that's just me...

I don't think `XElement` is useful as an interface...

**Comment by airbreather** [comment]
I'm unassigning myself from this one.  Sorry to bail out on this one after saying I'd take it on, but I'm not comfortable with attaching my name to any of the cheap designs that have been thrown around thus far, nor am I willing to invest the effort in my own time to make GPX I/O code that I think is worthy of being maintained in NTS itself.

In my opinion, the output of `xsd.exe /c GpxTypes.xsd` (which appears incompatible with .NET Standard 1.0 and minimally requires extra framework dependencies to work in .NET Standard 1.3) and some surface-level extension methods (the most useful of which initializes a bounding box based on the included points, if it's not initialized already) are not so expensive that it would be an unreasonable burden to expect an application which produces and consumes GPX data to maintain those things themselves.

> In order to write things to file that are relevant only to gpx one needs the gpx types, I don't see a way around it.

An alternative would be to convert the geometry data from GPX into GeoJSON format, with attributes, and then stuff the entire GeoJSON object(s) inside a containing object that stores the other GPX metadata.

> I can't relate to the statement that json is a dependency as I don't think there's an app that is written nowadays that does not use this package, but maybe that's just me...

Even if no such application exists, having to depend on NTS.IO.GeoJSON instead of "one NTS.IO package to rule them all" is not the end of the world.

> I don't think `XElement` is useful as an interface...

Perhaps not.  Given that GPX seems to be composed of distinct parts where a different subset may be relevant to any given use case, what I'd had in mind was something like [this](https://gist.github.com/airbreather/f374f8085af45ff185e539c0ae4f0024), where the GPX element could be part of a larger list of elements, and someone might want to just query for some subsection of the data in a beautiful XLINQ expression tree.  If XLINQ isn't as beautiful to you as it is to me, or if the use case is just a single `gpx` root element at a time and you don't need all that crap, `Gpx.Load(XDocument.Load(someFile).Root)` is not, in my opinion, less useful than `Gpx.Load(File.ReadAllBytes(someFile))`.

Obviously, I've only implemented a subset of the spec for this demo, and without GeoAPI types quite yet because I'm not sure what makes sense for it.

**Comment by HarelM** [comment]
Just want to clear a few things:
1. I'm not one of the maintainers of this project.
2. I don't think every application should generate the same gpx.cs file - If I'm using a framework (and NTS is a framework) I except common things to be in it, much like the fact that I don't re-implement `xmlserializer` in every project that needs it.
3. I think there should be one NTS.IO and not many packages, and that this package will only have things that are in use - GPX is, as far as my research went, the most popular file for recording a GPS and sharing geo related information, so it definitely deserves to be in this kind of framework, even more than GeoJson, IMHO.
4. The `UpdateBounds` is something I needed, I don't think it should be part of this framework although I don't think it has such a negative effect - it's extremely helpful to know that the bounds are always initialized.
5. The `toBytes` extension should not be in this project, IMO.

I did add the gpx code back to my project and it wasn't hard, but I believe other users might need it.
I don't think there should be a complicated design around reading and writing gpx files, feel to me like your are over complicating it, I can send a pull request if you don't feel up to it, although I do think you should finish what you started :-)
I think your original though was just fine:
> IMO, the furthest we should go is expose one method that reads a single GPX element from an active XmlReader provided by the caller (which might be inside some outer list, backed by a remote Stream, encoded in some unusual encoding), and the XmlWriter equivalent for producing the same.

Something like:
`void WriteGpx(gpxType gpx, XmlWriter writer)`
`gpxType ReadGpx(XmlReader reader)`
And maybe the above two with a `Stream` instead of writer/reader.
That's it...

As a side note, your implementation does not take into consideration the extension objects, also since the xsd generates a file I don't see a good reason to write a class that deserializes gpx as you would probably miss some of the pitfalls xsd is saving you from...

**Comment by DGuidi** [comment]
No bad feelings, thanks anyone for any kind of support. 
If there is a friction, i.e: some doubts about adding or maintaining the code, I assume the best way is to separate code in another repository.

**Comment by jaundice** [comment]
Hi Joe, just to clarify the existence of the cloudstorage.
It exists because I found it useful ☺ It allows multiple front end webservers (in my case) to access a single set of files on shared azure storage without having to download local copies

Regards,

John

**Comment by HarelM** [comment]
Here's my initial thoughts, let me know what you think be fore I dive deep:
Base the code to be similar to `ShapeReader/Writer`.
Some thing like:
```cs
    public class GpxReader
    {
        public IGeometry[] Read(Stream stream)
        public IGeometry[] Read(XmlReader reader)
    }
```
The problem here is that GPX has extra data that is not available for `IGeometry`, To solve this I thought about using `UserData` and place a `AttributesTable`. 
To use this, besides the casting, I thought about exposing constants for string names keys.

Let me know what you think.

**Comment by xivk** [comment]
Keep in mind for this issue that in the meantime we started to split off the IO.* packages into separate repos, so the best approach would be to start a new project and add NTS as a dependency. See #191 

@HarelM I can make a new repo under the NTS organization and give you access.

**Comment by HarelM** [comment]
My next sentence was going to be that I was hoping to merge all the io project that doesn't have a dependency with other packages such as shape, gpx, gdb.
I think the complexity of multiple repositories and dlls is not useful, IMHO.

**Comment by xivk** [comment]
It is very useful because of one main reason, release management. 

It's way easier to update and validate one package/project/dll and release a stable. Now we have to wait until NTS is released to update the IO code. It will make things move faster and new IO packages can be built and released faster. 

Think about TopoJSON for example, if someone decides to work on it now the only limit is quality and how fast it can be done.

On top of that it makes managing issue better and NTS is supposed to be in sync with JTS, the main repo can focus on that goal.

**Comment by xivk** [comment]
Other groupings can make sense for example all JSON-based formats (GeoJSON, TopoJSON) or XML-based formats (KML & GPX).

**Comment by HarelM** [comment]
No obejection in terms of a repo for NTS and a repo for IO but I wouldn't split IO into separate repositories - also management of releases can be made within the same repository.
I don't think the number of contributors, the rate of the delivery of these packages and the amount of code qualifies for repository split (not even for IO) as I think the overhead of working in separate repositories for the contributors team outweighs the advantages of repositories split, but it might be only me...
In any case, split for xml and json seems like a valid grouping.

**Comment by HarelM** [comment]
@xivk I think I have some time to work on this now, can you create a repository for me called NetTopologySuite.IO.XML (unless you prefer a different name).
If can configure the solution and project to make sure it compiles to all the relevant configurations it will be great, if not I can probably copy from another similar repository...

**Comment by HarelM** [comment]
@airbreather seems like a lot of work for reading/writing when everything is defined in an xsd file, but I might be missing something... let me know if you need any help.
Also please consider supporting extensions somehow.

**Comment by airbreather** [comment]
> seems like a lot of work for reading/writing when everything is defined in an xsd file, but I might be missing something

@HarelM some goals that I have:
1. It should be fast enough.
1. It should be memory-efficient; for the sake of running on low-spec mobile devices, I don't want the design, at least at this architectural level, to force a ton of junk to live in memory all at the same time.
1. It should produce idiomatic CLR types, even at the expense of perfect fidelity (which only matters in a real sense when you do a read/write round-trip and see surface-level discrepancies).

The `xsd.exe`-generated code, from my reading, **might** be fast enough (I'll give it the benefit of the doubt, since I haven't measured), but it fails to be memory-efficient (**everything** must be loaded before **anything** is returned to the caller, or else every caller who cares basically winds up doing major parts of what I've done here anyway), and it also has some flaws on the "idiomatic CLR types" side of things:
- `decimal`, everywhere.  Technically, this is more faithful to the source data and allows for exact round-tripping as long as there aren't too many digits of precision, but NTS and all modern computer hardware is optimized for `double`.
- Extensions are just `System.Xml.XmlElement[]`.  It just gives up trying to deserialize it, because it (by design) can't really do anything.  Now, while there's no getting around the fact that we can't completely solve the problem "in-the-box", we can still give the caller an opportunity to parse the XML however they want into, again, something more idiomatic (and vice-versa).  This only needs to happen once at the point of reading and once at the point of writing, for each element with extensions on it.
- There's no range validation, at least none that I can see.  You can totally try to save a waypoint with latitude >90, and there's nothing in there that will stop you until you go to write it (if anything even tries to stop you then!).
- Everything's mutable.  There are pros and cons to using mutable data types to represent this kind of stuff.  Personally, I favor having important payloads be immutable, for several reasons that I don't want to get into right now.
- Some other really odd type-related choices...
  - `wptType.sat` and `trkType.number` are `string` instead of `int` or `uint`
  - `linkType.href` is `string` instead of `System.Uri`
  - All optional value-type members seem to be generated as `foo` / `fooSpecified` pairs instead of just making `foo` nullable for parity with reference-type members (there can be some advantages to doing it this way, but those advantages only apply at the field level; property accessors can easily encapsulate this).  Some of those value-type members don't even need extra storage to be nullable at all, because they're range-restricted (either by definition or naturally)... so we can just pick some sentinel value outside the range and then say, e.g., "if the elevation equals one light-year, then that actually means that no elevation was specified, so return `null`" and save one byte per point.
  - Timestamps are just whatever `System.DateTime` reads, which is often going to be ambiguous and vary depending on what machine the software is running on.  There's no getting around **somebody** having to disambiguate to get usable information out of "`2018-07-17T08:15:00.000`", and IMO it makes the most sense to have the caller do so at the point where it's read, by saying "where the input data doesn't say a time zone, assume XYZ time zone" and then we can do that as we load the data.  Then, everybody who consumes the data will see a consistent, unambiguous timestamp.

So while I appreciate how easy it is to use `xsd.exe` to get "something" up and running, I don't see the generated code being a good foundation for something that I would personally want to distribute.

**Comment by airbreather** [comment]
@HarelM I've released version 0.1.0 [NetTopologySuite.IO.GPX](https://www.nuget.org/packages/NetTopologySuite.IO.GPX).  This API is very unstable and is subject to change multiple times before I find something I'm comfortable calling 1.x, as feedback comes in from users.

Pre-release packages are being automatically pushed to my personal MyGet feed as part of the Travis-CI build process: https://myget.org/feed/airbreather/package/nuget/NetTopologySuite.IO.GPX

I'll eventually work on some more documentation for it, but most of it should be discoverable through IntelliSense starting from these three classes:
- `NetTopologySuite.IO.GpxReader`
- `NetTopologySuite.IO.GpxWriter`
- `NetTopologySuite.IO.NetTopologySuiteGpxFeatureConverter`

If you find problems with it, please report them on [its own project site](https://github.com/NetTopologySuite/NetTopologySuite.IO.GPX).

I'm going to close this issue now that I have something to show on NuGet.

**Comment by HarelM** [comment]
@airbreather Thank! I'll give it a test run soon...
